test/testutils.cpp

```diff
@@ -483,6 +484,39 @@ class TestUtils : public TestFixture {
             ASSERT_EQUALS("test", *it);
         }
     }
+
+    void as_const() const {
+        struct C {
+            bool written = false;
+            void f() {
+                written = true;
+            }
+            void f() const {}
+        };
+
+        {
+            C c;
+            c.f();
+            ASSERT(c.written);
+        }
+        {
+            C c;
+            utils::as_const(c).f();
+            ASSERT(!c.written);
+        }
+        {
+            C c;
+            C* cp = &c;
+            cp->f();
+            ASSERT(c.written);
+        }
+        {
+            C c;
+            C* cp = &c;
+            utils::as_const(cp)->f(); // (correctly) calls non-const version
```

Comment:
That's why we need to dereference the pointers in the actual code.

I filed tickets about detecting this: https://github.com/llvm/llvm-project/issues/101580 / https://trac.cppcheck.net/ticket/12977.